Joss paper (#651)
FAIRmat-NFDI/pynxtools: paper/paper.md

    family-names: Márquez Prieto
    orcid: https://orcid.org/0000-0002-8173-2566
    affiliation: 1
  - given-names: Rubel
    family-names: Mozumder
    orcid: https://orcid.org/0009-0007-5926-6646
    affiliation: 1
  - given-names: Tommaso
    family-names: Pincelli
    orcid: https://orcid.org/0000-0003-2692-2540
    affiliation: 4
  - given-names: Walid
    family-names: Hetaba
    orcid: https://orcid.org/0000-0003-4728-0786
    affiliation: 3
  - given-names: Laurenz
    family-names: Rettig
    orcid: https://orcid.org/0000-0002-0725-6696
    affiliation: 4
  - given-names: Christoph
    family-names: Koch
    orcid: https://orcid.org/0000-0002-3984-1523
    affiliation: 1
  - given-names: Heiko B.
    family-names: Weber
    orcid: https://orcid.org/0000-0002-6403-9022
    affiliation: 2
  - given-names: Markus
    family-names: Scheidgen
    orcid: https://orcid.org/0000-0002-8038-2277
    affiliation: 1
  - given-names: Claudia
    family-names: Draxl
    orcid: https://orcid.org/0000-0003-3523-6657
    affiliation: 1
  - given-names: Sandor
    family-names: Brockhauser
    orcid: https://orcid.org/0000-0002-9700-4803
    affiliation: 1

affiliations:

>   - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Berlin, Germany

Reviewer: Full address required

    index: 1
  - name: Lehrstuhl für Angewandte Physik, Friedrich-Alexander-Universität Erlangen-Nürnberg, Erlangen, Germany
    index: 2
  - name: Department Heterogeneous Reactions, Max Planck Institute for Chemical Energy Conversion, Mülheim an der Ruhr, Germany
    index: 3
  - name: Department of Physical Chemistry, Fritz Haber Institute of the Max Planck Society, Berlin, DE
    index: 4
  - name: Felix-Bloch-Institut für Festkörperphysik, Universität Leipzig, Leipzig, Germany
    index: 5
date: 06 June 2025
bibliography: paper.bib

---

# Summary

> Scientific data across experimental physics and materials science often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022] due to incompatible instrument-specific formats and diverse standardization practices. pynxtools is a Python software development framework with a CLI interface that standardizes data conversion for scientific experiments in materials characterization to the NeXus format [@Koennecke:2015] across diverse scientific domains. NeXus uses NeXus application definitions as their data storage specifications. pynxtools provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across photoemission spectroscopy, electron microscopy [@Prestat:2025], atom probe tomography, optical spectroscopy, scanning probe microscopy, and X-ray diffraction. Through its modular plugin architecture, pynxtools provides maps for instrument-specific raw data, and electronic lab notebook metadata, to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Reviewer: We should format all usage of software tools differently, maybe ``pynxtools``

Reviewer 2: From the Joss example paper, software tools like pint and elasticsearch are written with ``, we should do the same. So e.g. 
```markdown
powered by `Elasticsearch`
```

For internal module names, shall we change them to italics? So instead of **Metainfo**, we write _Metainfo_?

# Statement of need

> Achieving FAIR (Findable, Accessible, Interoperable, and Reproducible) data principles in experimental physics and materials science requires consistent implementation of standardized data formats. While NeXus provides comprehensive data specifications for structured storage of scientific data, pynxtools simplifies the implementation process for developers and researchers by providing guided workflows and automated validation to ensure complete compliance. Existing tools [@Koennecke:2024; @Jemian:2025] provide solutions with individual capabilities, but none offers a comprehensive end-to-end workflow for proper NeXus adoption. pynxtools addresses this critical gap by providing an accessible framework that enforces complete NeXus application definition compliance through automated validation, detailed error reporting for missing required data points, and clear implementation pathways via configuration files and extensible plugins. This approach transforms NeXus from a complex specification into a practical solution, enabling researchers to achieve true data interoperability without deep technical expertise in the underlying standards.

Reviewer: The jump from "standardized data formats" to "While NeXus" is too large. If there are several cross-community standards we should mention these and then state like "one is NeXus". Then we can continue with the next sentence.

The sentence "While ..., pynxtools" to me is a contrast where "pynxtools simplifies" has been injected right in the middle but where I would rather expect first an acknowledging statement like 
"NeXus provides comprehensive data specifications for structured storage of scientific data" but followed by a statement what the challenges are namely that "Existing tools provide .... but for proper NeXus adoption". 

Then one can start with "pynxtools addresses and simplifies this ... , and clear implementation pathways ... and extensible plugins."

# Dataconverter and validation

The __dataconverter__ module forms the core of pynxtools, combining instrument output files and data from electronic lab notebooks into NeXus-compliant HDF5 files. The converter performs three key operations: reading experimental data through specialized readers, validating against NeXus application definitions to ensure compliance with existence, shape, and format constraints, and writing valid NeXus/HDF5 output files.

> The __dataconverter__ provides a command line interface (CLI) to produce NeXus files where users can use one of the built-in readers for generic functionality or technique-specific reader plugins distributed as separate Python packages.

Reviewer: This sentence is repeated in the NOMAD section #661

> For developers, the __dataconverter__ provides an abstract __reader__ class to create plugins that process experiment-specific file formats and try to fill the NeXus specification. A __Template__ object is passed to the __reader__ by the __dataconverter__ that acts like a form that has to be filled by the __reader__. __Template__ is a subclass of a regular Python __dict__ class. It keeps a similar interface to a __dict__ to help developers write code they are familiar with. It transparently ensures structural compliance with the selected NeXus application definition. __Template__ categorizes all data elements according to the three requirement levels used in NeXus: required, recommended, and optional.

Reviewer: We should check capitalization is the same for highlighted terms

Reviewer: Can the point on "It keeps ... they are familiar with" not be made shorter and pulled into the previous sentence.

Reviewer: "The __Template__ transparently ensures structural compliance with the selected NeXus application definition and categorizes all data elements according to the three requirement levels used in NeXus: required, recommended, and optional."

> This allows the __dataconverter__ to control what the reader plugins provide and validate this against the selected NeXus application definition. Once the reader plugin returns a __Template__ object, the __dataconverter__'s validation routine checks if all required fields in the NeXus application definition exist. This includes complex dependency relationships such as inheritance chains or required children within optional parent groups. The validation then collects all data entities that have a specification and ensures data integrity by verifying the data type, shape, and constraints set by the specification. The __dataconverter__'s validation routine reports an error if a required data entity is found to be invalid. It also emits warnings in the CLI for any data it invalidates or cannot match to a known specification, helping users and developers identify potential issues while practically building a NeXus file.

Reviewer: "data entities" and "data elements" why variation?

> After validation, the __dataconverter__ passes the __Template__ object to the __Writer__ class. This class has the job of writing an HDF5 file as provided in a __Template__ object. It also converts the Template object's simplified link syntax into an HDF5 link, which may be a soft or hard link, either within the same file or across files. This link syntax is reduced to a dict object with a link key, where the value is a string in the format "<filename.hdf5>:<\/path\/to\/data\/in\/template\/object>". The filename is optional and can be ommitted if the data being linked is in the same file. This approach significantly simplifies linking to data for new users and developers. The same syntax is available in our dynamic mapping built-in readers.

Reviewer 2: I think it is fine for now, we are around a thousand words.

All reader implementations are validated through a test suite that ensures compatibility with the __dataconverter__ framework. We provide a set of tests under __pynxtools.testing__ for every plugin to use. These tests run automatically via the continuous integration pipeline provided by GitHub, maintaining code quality and functional integration across all reader plugins.

> The __dataconverter__ module is also shipped with an ELN generator that, for a selected NeXus application definition, creates either a YAML file that can be manually filled in and used by __dataconverter__ readers or a NOMAD ELN schema that can be used to manually fill in data on the NOMAD research data management platform [@Scheidgen:2023].

Reviewer: Reference to NOMAD upon first mentioning and spelling out of the acronym Novel Mat...

Author: I'm not so keen on that. Everyone knows it as NOMAD.

Reviewer 2: Agreed

# NeXus reader and annotator (read\_nexus)

__read_nexus__ is a command line tool to annotate and explore NeXus data files. Once pynxtools is installed, it is immediately available and offers options for selecting a NeXus file and either documenting all or part of its content. Note that with no specific NeXus file provided as an argument, the tool will use by default an example NeXus file shipped together with the package.

> _annotations -_ As it is expained above[!!! Ref better when it is done], NeXus defines a rich set of concepts, the so called [NeXus Vocabulary](https://manual.nexusformat.org/classes/index.html#classes-vocabulary-downloads). These include _Attributes_ and _Fields_ which are organized in _Groups_ and their subgroups. [NeXus files](https://manual.nexusformat.org/design.html) store their [data items](see details in https://manual.nexusformat.org/datarules.html) following the hierarchy of these groups. _read\_nexus_ traverses the input NeXus file and interprets each data item by following its group path and connecting it to the concept of the specific vocabulary item. It reports the concept's definition. Since NeXus concepts follow inheritance, the all superclass concepts along the inheritance chain and their definitions are also reported. If a data item in a NeXus file does not belong to any concept (which is allowed by the NeXus standard), _read\_nexus_ reports the message 'NOT IN SCHEMA'.

Reviewer 2: Was heavily shortened, so this comment does not apply anymore.

__semantic use__: _read\_nexus_ also supports the proper semantic use of data items in NeXus files by automating the annotation process of connecting them to the corresponding NeXus concepts. Hence, data processing applications can find data items in a NeXus file based on the semantic concepts, without the need of hardcoding any data path which could lead to incompatibilities. Note that _read\_nexus_ uses [NeXus Ontology's](https://github.com/nexusformat/NeXusOntology) [OWL](https://www.w3.org/OWL/) labels assigned to each NeXus' semantic concept. Hence, it can also be used to connect data items from any NeXus files to the semantic web. The following features facilitate such use:

_--concept_: This command line queries the NeXus file using a specific NeXus vocabulary item (or NeXus Ontology class), reporting all data item paths that correspond to this concept.

> _--documentation_: Instead of traversing the whole file, this feature annotates only a single data item at a specific path in the NeXus file. This is practical when only a few specific data items need to be investigated, e.g. those selected by a _--concept_ query.

Reviewer 2: I suggest (if possible) we keep it just a bit longer and make a full stop after "The tool identifies default plottable data according to NXdata specifications."

Then, we can continue by saying 

"It enables semantic querying of data items in NeXus files through the command line option --concept, which finds all data paths matching a given concept. --documentation finds the concept and documentation for a specific data path, allowing data processing applications to locate items by concept rather than hardcoded paths."

> _--documentation_: Instead of traversing the whole file, this feature annotates only a single data item at a specific path in the NeXus file. This is practical when only a few specific data items need to be investigated, e.g. those selected by a _--concept_ query.

Reviewer: "--documentation" > "The CLI option --documentation"

- Schema Package: The NeXus (meta)data definitions are expressed in XML using the NeXus Definition Language (NXDL), which in turn is defined using XSD. pynxtools converts this representation and extends NOMAD's internal data schema (called __Metainfo__) with these domain-specific quantities. pynxtools also connects the NeXus vocabulary to existing base sections in NOMAD — reusable, standardized building blocks used to represent common scientific concepts. This connection enables interoperability between NeXus-defined concepts and other standardized representations in NOMAD, such as those for sample synthesis or theoretical calculations.

- Data Converter: The __DataConverter__ as described above is also available in NOMAD. Thus, NOMAD users can directly convert their experimental data to NeXus using NOMAD's graphical interface. In addition to the capabilities already described, the internal __DataConverter__ class also handles NOMAD's electronic lab notebooks (ELNs) and converts these such that the manually inputted data can be converted to NeXus as well.

> - Parser: The NOMAD parser module in pynxtools (__NexusParser__) reads NeXus HDF5 files and uses the structured data instances from these files to populate the NOMAD __Metainfo__ model with __Metainfo__ object instances as defined by the pynxtools schema package. This step enables ingestion of NeXus data directly into the NOMAD __Metainfo__ model.

Reviewer 2: Yeah, please go ahead.

> - Parser: The NOMAD parser module in pynxtools (__NexusParser__) reads NeXus HDF5 files and uses the structured data instances from these files to populate the NOMAD __Metainfo__ model with __Metainfo__ object instances as defined by the pynxtools schema package. This step enables ingestion of NeXus data directly into the NOMAD __Metainfo__ model.

Reviewer 2: See #661

- App: pynxtools contains an integrated search application for NeXus data within NOMAD. This application, powered by Elasticsearch [@elasticsearch:2025], enables users to efficiently filter uploaded data based on various parameters, such as experiment type, upload timestamp, and other relevant quantities.

- Example Upload: The plugin includes a representative NOMAD upload (based on the NeXus application definition __NX_iv_temp__ describing temperature-dependent IV curve measurements), which exemplifies the entire workflow of pynxtools as a NOMAD package. This example upload details the conversion of data from experiments into NeXus files using the __DataConverter__, along with parsing them into the NOMAD archive. This example upload is designed for new users to understand the pynxtools workflow in NOMAD and serves as templates to adapt the plugin to new NeXus applications.

> 

Author: This can be reduced or removed. It's nice to know about. But this should already be something well mentioned in our documentation. It doesn't explain any feature.

Reviewer 2: See #661, to be discussed there

---

> title: 'pynxtools: A framework for generating NeXus files from formats across disciplines'

Reviewer: title: "A flexible Python framework for generating and verifying interoperable NeXus files across disciplines"

> title: 'pynxtools: A framework for generating NeXus files from formats across disciplines.'

Reviewer 2: Remove the period at the end

  - name: Physics Department and CSMB, Humboldt-Universität zu Berlin, Zum Großen Windkanal 2, D-12489 Berlin, Germany
    index: 1
    ror: 01hcx6992
  - name: Department Heterogeneous Reactions, Max Planck Institute for Chemical Energy Conversion, Stiftstraße 34-36, D-45470 Mülheim an der Ruhr, Germany
    index: 2
    ror: 01y9arx16
  - name: Department of Physics, RPTU Kaiserslautern-Landau, Erwin-Schrödinger-Str. 46, D-67663 Kaiserslautern, Germany
    index: 3
    ror: 01qrts582
  - name: Felix Bloch Institute for Solid State Physics, Leipzig University, Linnestr. 5, D-04103 Leipzig, Germany
    index: 4
    ror: 03s7gtk40
  - name: Department of Physical Chemistry, Fritz Haber Institute of the Max Planck Society, Faradayweg 4-6, D-14195 Berlin, DE
    index: 5
    ror: 03k9qs827
  - name: Institute of Physics, Chinese Academy of Sciences, No.8, 3rd South Street, Zhongguancun, Haidian District, Beijing, China
    index: 6
    ror: 05cvf7v30
  - name: Lehrstuhl für Angewandte Physik, Friedrich-Alexander-Universität Erlangen-Nürnberg, Staudtstr. 7, D-91058 Erlangen, Germany
    index: 7
    ror: 00f7hpc57

date: 07 July 2025
bibliography: paper.bib

---

# Summary

> Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. pynxtools is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials characterization to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus uses NeXus application definitions as their data storage specifications. pynxtools provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, X-ray diffraction. Through its modular plugin architecture, pynxtools provides maps for instrument-specific raw data, and electronic lab notebook metadata, to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Reviewer 2: "NeXus uses NeXus application definitions as their data storage specifications" 

This reads akwardly and doesn't explain what appdefs are. How about

"NeXus defines data storage specifications for different experimental techniques through application definitions"

> Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. pynxtools is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials characterization to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus uses NeXus application definitions as their data storage specifications. pynxtools provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, X-ray diffraction. Through its modular plugin architecture, pynxtools provides maps for instrument-specific raw data, and electronic lab notebook metadata, to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Reviewer 2: "scanning probe microscopy, X-ray diffraction."

connect with an "and"

> Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. pynxtools is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials characterization to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus uses NeXus application definitions as their data storage specifications. pynxtools provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, X-ray diffraction. Through its modular plugin architecture, pynxtools provides maps for instrument-specific raw data, and electronic lab notebook metadata, to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Reviewer 2: "Through its modular plugin architecture, pynxtools provides maps for instrument-specific raw data, and electronic lab notebook metadata, to these unified definitions"

These commas don't work here, you can just remove them

> Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. `pynxtools` is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials characterization to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus defines data storage specifications for different experimental techniques through application definitions. `pynxtools` provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, and X-ray diffraction. Through its modular plugin architecture `pynxtools` provides maps for instrument-specific raw data and electronic lab notebook metadata to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Author: ```suggestion
Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. `pynxtools` is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials science to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus defines data storage specifications for different experimental techniques through application definitions. `pynxtools` enables rapid developing and designing of these application definitions. It also provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, and X-ray diffraction. Through its modular plugin architecture `pynxtools` provides simple integration of instrument-specific raw data and electronic lab notebook metadata to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.
```

> Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. `pynxtools` is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials characterization to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus defines data storage specifications for different experimental techniques through application definitions. `pynxtools` provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, and X-ray diffraction. Through its modular plugin architecture `pynxtools` provides maps for instrument-specific raw data and electronic lab notebook metadata to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Reviewer: RDM should be mentioned in the abstract

NOMAD must not be mentioned in the title

> Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. `pynxtools` is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials science to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus defines data storage specifications for different experimental techniques through application definitions. `pynxtools` enables rapid developing and designing of these application definitions. It also provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, and X-ray diffraction. Through its modular plugin architecture `pynxtools` provides simple integration of instrument-specific raw data and electronic lab notebook metadata to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. By simplifying the adoption of standardized application definitions, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.

Reviewer 2: ```suggestion
Scientific data across physics, materials science, and materials engineering often lacks adherence to FAIR principles [@Wilkinson:2016; @Jacobsen:2020; @Barker:2022; @Wilkinson:2025] due to incompatible instrument-specific formats and diverse standardization practices. `pynxtools` is a Python software development framework with a command line interface (CLI) that standardizes data conversion for scientific experiments in materials science to the NeXus format [@Koennecke:2015; @Koennecke:2006; @Klosowski:1997] across diverse scientific domains. NeXus defines data storage specifications for different experimental techniques through application definitions.  `pynxtools` provides a fixed, versioned set of NeXus application definitions that ensures convergence and alignment in data specifications across, among others, atom probe tomography, electron microscopy, optical spectroscopy, photoemission spectroscopy, scanning probe microscopy, and X-ray diffraction. Through its modular plugin architecture `pynxtools` provides conversion of data and metadata from instruments and electronic lab notebooks to these unified definitions, while performing validation to ensure data correctness and NeXus compliance. `pynxtools` can be integrated directly into Research Data Management Systems (RDMS) to facilitate parsing and normalization. We detail one example for the RDM system NOMAD. By simplifying the adoption of NeXus, the framework enables true data interoperability and FAIR data management across multiple experimental techniques.
```

- removed "simple" because we repeat and specify this later
- removed sentence on "enables rapid developing and designing of these application definitions" -> this is not what this software is about
- added sentence on NOMAD integration -> substantial part of the software, important for automated searches of abstracts

# Statement of need

Achieving FAIR (Findable, Accessible, Interoperable, and Reproducible) data principles in experimental physics and materials science requires consistent implementation of standardized data formats. NeXus provides comprehensive data specifications for structured storage of scientific data. pynxtools simplifies the use of NeXus for developers and researchers by providing guided workflows and automated validation to ensure complete compliance. Existing tools [@Koennecke:2024; @Jemian:2025] provide solutions with individual capabilities, but none offers a comprehensive end-to-end workflow for proper NeXus adoption. pynxtools addresses this critical gap by providing a framework that enforces complete NeXus application definition compliance through automated validation, detailed error reporting for missing required data points, and clear implementation pathways via configuration files and extensible plugins. This approach transforms NeXus from a complex specification into a practical solution, enabling researchers to achieve true data interoperability without deep technical expertise in the underlying standards.

# Dataconverter and validation

> The _dataconverter_, core module of pynxtools, combines instrument output files and data from electronic lab notebooks into NeXus-compliant HDF5 files. The converter performs three key operations: reading experimental data through specialized readers, validating against NeXus application definitions to ensure compliance with existence, shape, and format constraints, and writing valid NeXus/HDF5 output files.

Reviewer 2: ```suggestion
The `dataconverter`, core module of pynxtools, combines instrument output files and data from electronic lab notebooks into NeXus-compliant HDF5 files. The converter performs three key operations: extracting experimental data through specialized readers, validating against NeXus application definitions to ensure compliance with existence, shape, and format constraints, and writing valid NeXus/HDF5 output files.
```

> The `dataconverter`, core module of pynxtools, combines instrument output files and data from electronic lab notebooks into NeXus-compliant HDF5 files. The converter performs three key operations: extracting experimental data through specialized readers, validating against NeXus application definitions to ensure compliance with existence, shape, and format constraints, and writing valid NeXus/HDF5 output files.

Reviewer: Yes we should, symbols system also has active discussions within the NIAC currently, it is "essentially" broken and an active point of discussion there.

The _dataconverter_ provides a CLI to produce NeXus files where users can use one of the built-in readers for generic functionality or technique-specific reader plugins, distributed as separate Python packages.

For developers, the _dataconverter_ provides an abstract _reader_ class for building plugins that process experiment-specific formats and populate the NeXus specification. It passes a _Template_, a subclass of Python’s dict, to the _reader_ as a form to fill. The _Template_ ensures structural compliance with the chosen NeXus application definition and organizes data by NeXus's required, recommended, and optional levels.

> The _dataconverter_ validates _reader_ output against the selected NeXus application definition, checking required fields, complex dependencies (like inheritance and nested group rules), and data integrity (type, shape, constraints). It reports errors for invalid required fields and emits CLI warnings for unmatched or invalid data, aiding practical NeXus file creation.

Reviewer 2: ```suggestion
The `dataconverter` validates `reader` output against the selected NeXus application definition, checking required fields, complex dependencies (like inheritance and nested group rules), and data integrity (type, shape, constraints). It reports errors for invalid required fields and emits CLI warnings for unmatched or invalid data, aiding practical NeXus file creation.
```

All reader plugins are tested using the pynxtools.testing suite, which runs automatically via GitHub CI to ensure compatibility with the dataconverter, the NeXus specification, and integration across plugins.

> The dataconverter includes an ELN generator that creates either a fillable `YAML` file or a `NOMAD` [@Scheidgen:2023] ELN schema based on a selected NeXus application definition.

Reviewer 2: This is this tool: https://github.com/FAIRmat-NFDI/pynxtools/tree/master/src/pynxtools/eln_mapper

It was originally developed by @RubelMozumder and then I made significant updates. I agree, we need (better) documentation for this.

# NeXus reader and annotator

> _read_nexus_ enables semantic access to NeXus files by linking data items to NeXus concepts, allowing applications to locate relevant data without hardcoding file paths. It supports concept-based queries that return all data items associated with a specific NeXus Vocabulary term. Each data item is annotated by traversing its group path and resolving its corresponding NeXus concept, included inherited definitions.

Reviewer 2: ```suggestion
`read_nexus` enables semantic access to NeXus files by linking data items to NeXus concepts, allowing applications to locate relevant data without hardcoding file paths. It supports concept-based queries that return all data items associated with a specific NeXus Vocabulary term. Each data item is annotated by traversing its group path and resolving its corresponding NeXus concept, included inherited definitions.
```

Items not part of the NeXus schema are explicitly marked as such, aiding in validation and debugging. Targeted documentation of individual data items is supported through path-specific annotation. The tool also identifies and summarizes the file’s default plottable data based on the NXdata definition.

# `NOMAD` integration

> While pynxtools works as a standalone tool, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):

Reviewer 2: If we write `NOMAD`, we should probably write `pynxtools`

> While pynxtools works as a standalone tool, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):

Reviewer 2: If we write `NOMAD`, we should probably write `pynxtools`

> While pynxtools works as a standalone tool, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):

Reviewer 2: ```suggestion
While `pynxtools` works as a standalone tool, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):
```

> While `pynxtools` works as a standalone tool, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):

Reviewer: ```suggestion
While `pynxtools` works independently, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):
```

> While `pynxtools` works as a standalone tool, it can also be integrated directly into Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):

Reviewer: ```suggestion
While `pynxtools` works as a standalone tool, it can be integrated to operate with Research Data Management Systems (RDMS). Out of the box, the package functions as a plugin within the `NOMAD` platform [@Scheidgen:2023; @Draxl:2019]. This enables data in the NeXus format to be integrated into `NOMAD`'s metadata model, making it searchable and interoperable with other data from theory and experiment. The plugin consists of several key components (so called entry points):
```

> pynxtools extends `NOMAD`'s data schema (called _Metainfo_ [@Ghiringhelli:2017]) by integrating NeXus definitions as a `NOMAD` Schema Package, adding NeXus-specific quantities and enabling interoperability through links to other standardized data representations in `NOMAD`. The _dataconverter_ is integrated into `NOMAD`, making the conversion of data to NeXus accessible via the `NOMAD` GUI. The _dataconverter_ also processes manually entered `NOMAD` ELN data in the conversion.

Reviewer 2: ```suggestion
`pynxtools` extends `NOMAD`'s data schema (called `Metainfo` [@Ghiringhelli:2017]) by integrating NeXus definitions as a `NOMAD` _Schema Package_, adding NeXus-specific quantities and enabling interoperability through links to other standardized data representations in `NOMAD`. The `dataconverter` is integrated into `NOMAD`, making the conversion of data to NeXus accessible via the `NOMAD` GUI. The `dataconverter` also processes manually entered `NOMAD` ELN data in the conversion.
```

> The `NOMAD` Parser module in pynxtools (_NexusParser_) extracts structured data from NeXus HDF5 files to populate `NOMAD` with _Metainfo_ object instances as defined by the pynxtools schema package. This enables ingestion of NeXus data directly into `NOMAD`. Parsed data is post-processed using `NOMAD`'s Normalization pipeline. This includes automatic handling of units, linking references (including sample and instrument identifiers defined elsewhere in `NOMAD`), and populating derived quantities needed for advanced search and visualization.

Reviewer 2: ```suggestion
The `NOMAD` Parser module in `pynxtools` (`NexusParser`) extracts structured data from NeXus HDF5 files to populate `NOMAD` with `Metainfo` object instances as defined by the `pynxtools` schema package. This enables ingestion of NeXus data directly into `NOMAD`. Parsed data is post-processed using `NOMAD`'s _Normalization_ pipeline. This includes automatic handling of units, linking references (including sample and instrument identifiers defined elsewhere in `NOMAD`), and populating derived quantities needed for advanced search and visualization.
```

pynxtools contains an integrated Search Application for NeXus data within `NOMAD`, powered by `Elasticsearch` [@elasticsearch:2025]. This provides a search dashboard whereby users can efficiently filter uploaded data based on parameters like experiment type, upload timestamp, and domain- and technique-specific quantities. The entire pynxtools workflow (conversion, parsing, and normalization) is exemplified in a representative `NOMAD` Example Upload that is shipped with the package. This example helps new users understand the workflow and serves as a template to adapt the plugin to new NeXus applications.

# Funding
The work is funded by the Deutsche Forschungsgemeinschaft (DFG, German Research Foundation) - 460197019 (FAIRmat).

# Acknowledgements

> We acknowledge the following software packages our package depends on: [@H5py:2008], [@Harris:2020], [@Click:2014], [@Druskat:2021], [@Hoyer:2017], [@Hoyer:2025], [@Pandas:2020], [@McKinney:2010], [@Behnel:2005], [@Clarke:2019], [@Hjorth:2017], [@Pint:2012].

Reviewer 2: These didn't look good in the PDF. 
<img width="1038" height="142" alt="image" src="https://github.com/user-attachments/assets/a53bada4-bc12-4aa3-b938-14dc220acf5c" />

Maybe change to
```
"`software-in-math` [citation]" 
```

For example,
```
"`click` [@Click:2014]" 
```

H5py authors needs to be displayed correctly as well.

# Research Impact

`pynxtools` is integrated into production systems, demonstrating real-world impact across multiple deployment contexts. It is designed as a plugin for `NOMAD` to support NeXus and the features that come with our framework. Within the `NOMAD` platform, it underpins NeXus-based ingestion, validation, and search of experimental datasets, enabling automatic unit handling, metadata linking, and advanced search over NeXus-defined quantities. The framework is also distributed with `NOMAD` Oasis installations operated in experimental laboratories, demonstrating its suitability for decentralized, on-premise data management.

> The `NOMAD` integration consists of several key components: `pynxtools` extends `NOMAD`'s data schema (`Metainfo` [@Ghiringhelli:2017]) by integrating NeXus definitions as a `Schema Package`, introducing NeXus-specific quantities and enabling interoperability with other standardized data representations. The `NexusParser` module extracts structured data from NeXus HDF5 files to populate `NOMAD` with `Metainfo` object instances. Parsed data is post-processed using `NOMAD`'s `Normalization` pipeline, including automatic handling of units, linking references (sample and instrument identifiers), and populating derived quantities needed for advanced search and visualization. An integrated `Search Application` powered by `Elasticsearch` [@elasticsearch:2025] provides a dashboard for filtering uploaded data based on experiment type, upload timestamp, and domain-specific quantities. The entire workflow is exemplified in a representative `Example Upload` shipped with the package.

Reviewer 2: ```suggestion
The `NOMAD` integration consists of several key components: the `NOMAD`'s data schema (`Metainfo` [@Ghiringhelli:2017]) is extended by integrating NeXus definitions as a `Schema Package`, introducing NeXus-specific quantities and enabling interoperability with other standardized data representations. The `NexusParser` module within `pynxtools` extracts structured data from NeXus HDF5 files to populate `NOMAD` with `Metainfo` object instances. Parsed data is post-processed using `NOMAD`'s `Normalization` pipeline, including automatic handling of units, linking references (sample and instrument identifiers), and populating derived quantities needed for advanced search and visualization. An integrated `Search Application` powered by `Elasticsearch` [@elasticsearch:2025] provides a dashboard for filtering uploaded data based on experiment type, upload timestamp, and domain-specific quantities. The entire workflow is exemplified in a representative `Example Upload` shipped with the package.
```

Minimal change to indicate that the schema package may still move into `nomad-measurements`, but `pynxtools` still generates and uses the schema.

---

> title: 'Pynxtools: A framework for generating NeXus files from raw file formats.'

Reviewer: What about the following terms and thoughts in the title
- pynxtools, I agree,
- NeXus, I agree
- "across disciplines" is one of our USP
- semantic harmonization is another of our USP
- "raw" file formats, I disagree can be used but the term is ambiguous, most file formats store not only raw data

Reviewer 3: pynxtools : A Software Toolbox for Generating and Working wirh NeXus Files

tags:
  - Python
  - nexus
  - data modeling
  - file format

>   - data parsing

Reviewer: add "data normalization"
add "serialization"
add "HDF5"

(Missing what NeXus is. Add a bit of context.)
(Add it's a CLI tool. And code framework)
(Nexus Reference, https://doi.org/10.1107/S1600576714027575)
(Compromise for Nomad: We have a standalone tool that is also integrated in Nomad as an example.)
(https://github.com/ess-dmsc/nexus-constructor)

> (Cite some FAIR papers, one from Jacobson, FAIR4RS, FAIR4Workflows-thisnotsoimp)

Reviewer: Jacobsen: https://doi.org/10.1162/dint_r_00024
FAIR4RS: https://doi.org/10.1038/s41597-022-01710-x
NeXus constructor: https://doi.org/10.1088/1748-0221/13/10/T10001
NeXus reference: https://doi.org/10.1016/S0921-4526(97)00865-X, and the link you have above already

# Statement of need

> Scientific data across experimental physics and materials science remains largely non-FAIR (Findable, Accessible, Interoperable, and Reproducible) due to inconsistent implementation and documentation of standardized data formats. While NeXus provides comprehensive data specifications for structured scientific data storage, researchers typically struggle with its specification requirements, leading to incomplete implementations, non-compliant outputs, or abandonment of standardization efforts entirely. Existing tools (do some parts of what we offer, and then list what we provide) lack robust validation frameworks and provide insufficient guidance for proper NeXus adoption. pynxtools addresses this critical gap by providing an accessible framework that enforces complete NeXus application definition compliance through automated validation, detailed error reporting for missing required data points, and clear implementation pathways via configuration files and extensible plugins. This approach transforms NeXus from a complex specification into a practical solution, enabling researchers to achieve true data interoperability without deep technical expertise in the underlying standards.

Reviewer: "condensed-matter physics" not experimental physics only

> Scientific data across experimental physics and materials science remains largely non-FAIR (Findable, Accessible, Interoperable, and Reproducible) due to inconsistent implementation and documentation of standardized data formats. While NeXus provides comprehensive data specifications for structured scientific data storage, researchers typically struggle with its specification requirements, leading to incomplete implementations, non-compliant outputs, or abandonment of standardization efforts entirely. Existing tools (do some parts of what we offer, and then list what we provide) lack robust validation frameworks and provide insufficient guidance for proper NeXus adoption. pynxtools addresses this critical gap by providing an accessible framework that enforces complete NeXus application definition compliance through automated validation, detailed error reporting for missing required data points, and clear implementation pathways via configuration files and extensible plugins. This approach transforms NeXus from a complex specification into a practical solution, enabling researchers to achieve true data interoperability without deep technical expertise in the underlying standards.

Reviewer: The point I raised was, we can reformulate like:
The transitioning from paper to electronic lab notebooks paired with efforts to report research as structured data rather than prose (according to the FAIR principles as lead to an increased use of semantic modeling (including above-mentioned our target communities, i.e. condensed matter, materials science, as well as materials engineering). These essentially build domain-specific ontologies (adds refs) along the semantic web technology stack. Achieving true interoperability though, calls for a mapping of concepts to enable the implementation of software tools for asserting if two concepts are equivalent or different (and in which regards). Different strategies exist to implement such mapping: Explicit mapping rules (e.g. SSSOM). A complementary alternative is standardization, via negotiating of concensus across multiple research communities and subsequently agreeing on a shared set of overall fewer  semantic concepts. This strategy is used by NeXus.

(Can we tone down the first sentence to not offend readers/people who worked before. Don't dwell on stuff not working. How to make this sound less negative?)
(Second sentence: We base everything on NeXus. We don't wanna say it's too complicated. Make it nicer. We make it more easier to handle it.)

> (Third: refer to cnxvalidate and such)

Reviewer: Is fixed in https://github.com/FAIRmat-NFDI/pynxtools/commit/473e2831bfcd4386f2b577ad8725b02b862fb5c9
